File: app/routes/data_entries.py
from flask import Blueprint, request, jsonify
from app.models import Company, DataEntry
from app.database import db
from sqlalchemy import func

bp = Blueprint('data_entries', __name__, url_prefix='')


# GET all data entries (optionally filtered)
@bp.route('/data-entries', methods=['GET'])
def get_data_entries():
    company_name = request.args.get('company_name')
    uid = request.args.get('uid')
    data_set = request.args.get('data_set')

    query = DataEntry.query.join(Company)

    if company_name:
        query = query.filter(Company.name == company_name)
    if uid:
        query = query.filter(DataEntry.uid == uid)
    if data_set:
        query = query.filter(DataEntry.data_set == data_set)

    data_entries = query.all()
    return jsonify([entry.to_dict() for entry in data_entries])


# POST a new data entry
@bp.route('/data-entries', methods=['POST'])
def create_data_entry():
    data = request.get_json()

    if not data or 'company_id' not in data:
        return jsonify({'error': 'company_id is required'}), 400

    company = Company.query.get(data['company_id'])
    if not company:
        return jsonify({'error': 'Company not found'}), 404

    data_entry = DataEntry(
        company_id=data['company_id'],
        device_type=data.get('device_type'),
        uid=data.get('uid'),
        data_type=data.get('data_type'),
        data_set=data.get('data_set'),
        data_going_to=data.get('data_going_to')
    )

    db.session.add(data_entry)
    db.session.commit()

    return jsonify(data_entry.to_dict()), 201


# GET a specific data entry
@bp.route('/data-entries/<int:entry_id>', methods=['GET'])
def get_data_entry(entry_id):
    data_entry = DataEntry.query.get_or_404(entry_id)
    return jsonify(data_entry.to_dict())


# PUT update a data entry
@bp.route('/data-entries/<int:entry_id>', methods=['PUT'])
def update_data_entry(entry_id):
    data_entry = DataEntry.query.get_or_404(entry_id)
    data = request.get_json()

    if not data:
        return jsonify({'error': 'No data provided'}), 400

    if 'device_type' in data:
        data_entry.device_type = data['device_type']
    if 'uid' in data:
        data_entry.uid = data['uid']
    if 'data_type' in data:
        data_entry.data_type = data['data_type']
    if 'data_set' in data:
        data_entry.data_set = data['data_set']
    if 'data_going_to' in data:
        data_entry.data_going_to = data['data_going_to']

    db.session.commit()
    return jsonify(data_entry.to_dict())


# DELETE a data entry
@bp.route('/data-entries/<int:entry_id>', methods=['DELETE'])
def delete_data_entry(entry_id):
    data_entry = DataEntry.query.get_or_404(entry_id)
    db.session.delete(data_entry)
    db.session.commit()
    return jsonify({'message': 'Data entry deleted successfully'})


# GET stats for a specific company
@bp.route('/stats/company/<int:company_id>', methods=['GET'])
def get_company_stats(company_id):
    company = Company.query.get_or_404(company_id)

    total_entries = DataEntry.query.filter_by(company_id=company_id).count()

    data_set_counts = db.session.query(
        DataEntry.data_set,
        func.count(DataEntry.id).label('count')
    ).filter_by(company_id=company_id).group_by(DataEntry.data_set).all()

    device_type_counts = db.session.query(
        DataEntry.device_type,
        func.count(DataEntry.id).label('count')
    ).filter_by(company_id=company_id).group_by(DataEntry.device_type).all()

    return jsonify({
        'company': company.to_dict(),
        'total_entries': total_entries,
        'data_set_counts': [{'data_set': ds, 'count': count} for ds, count in data_set_counts],
        'device_type_counts': [{'device_type': dt, 'count': count} for dt, count in device_type_counts]
    })


# GET count of entries by company name and data_set
@bp.route('/stats/data-set-count', methods=['GET'])
def get_data_set_count():
    company_name = request.args.get('company_name')
    data_set = request.args.get('data_set')

    if not company_name or not data_set:
        return jsonify({'error': 'company_name and data_set parameters are required'}), 400

    count = db.session.query(func.count(DataEntry.id)).join(Company).filter(
        Company.name == company_name,
        DataEntry.data_set == data_set
    ).scalar()

    return jsonify({
        'company_name': company_name,
        'data_set': data_set,
        'count': count
    })

File: app/static/js/data_entries.js
// data_entries.js - Data entries related functions

import { dataEntries, showMessage, apiRequest } from './main.js';

function openEntryModal() {
    document.getElementById('entry-modal').classList.remove('hidden');
}

function closeEntryModal() {
    document.getElementById('entry-modal').classList.add('hidden');
}


// Load companies for the filter dropdown in Data Entries tab
async function loadCompaniesForDataFilters() {
    try {
        const companies = await apiRequest('/companies');
        const filterSelect = document.getElementById('filter-company');

        filterSelect.innerHTML = '<option value="">All Companies</option>';

        companies.forEach(company => {
            const option = document.createElement('option');
            option.value = company.id;
            option.textContent = company.name;
            filterSelect.appendChild(option);
        });

        console.log(`Loaded ${companies.length} companies into filter dropdown`);
    } catch (error) {
        console.error('Failed to load filter companies:', error);
        showMessage('data-message', 'Failed to load companies for filtering', 'error');
    }
}


// Load data entries from API
async function loadDataEntries() {
    try {
        const params = new URLSearchParams();
        
        const companyName = document.getElementById('filter-company')?.value;
        const uid = document.getElementById('filter-uid')?.value;
        const dataSet = document.getElementById('filter-data-set')?.value;
        
        if (companyName) params.append('company_name', companyName);
        if (uid) params.append('uid', uid);
        if (dataSet) params.append('data_set', dataSet);

        const queryString = params.toString();
        const endpoint = `/data-entries${queryString ? '?' + queryString : ''}`;
        
        const entriesData = await apiRequest(endpoint);
        // Update global dataEntries array
        dataEntries.length = 0; // Clear existing
        dataEntries.push(...entriesData); // Add new data
        renderDataEntries();
    } catch (error) {
        showMessage('data-message', 'Failed to load data entries', 'error');
    }
}

// Render data entries table
function renderDataEntries() {
    const container = document.getElementById('data-entries-list');
    
    if (!container) {
        console.warn('data-entries-list container not found');
        return;
    }
    
    if (dataEntries.length === 0) {
        container.innerHTML = '<div class="loading">No data entries found</div>';
        return;
    }

    let html = `
        <div class="table-container">
            <table>
                <thead>
                    <tr>
                        <th>ID</th>
                        <th>Company</th>
                        <th>Device Type</th>
                        <th>UID</th>
                        <th>Data Type</th>
                        <th>Data Set</th>
                        <th>Data Going To</th>
                        <th>Actions</th>
                    </tr>
                </thead>
                <tbody>
    `;

    dataEntries.forEach(entry => {
        html += `
            <tr>
                <td>${entry.id}</td>
                <td>${entry.company_name || 'N/A'}</td>
                <td>${entry.device_type || 'N/A'}</td>
                <td>${entry.uid || 'N/A'}</td>
                <td>${entry.data_type || 'N/A'}</td>
                <td>${entry.data_set || 'N/A'}</td>
                <td>${entry.data_going_to || 'N/A'}</td>
                <td>
                    <button class="action-btn delete-btn" onclick="deleteDataEntry(${entry.id})">
                        Delete
                    </button>
                </td>
            </tr>
        `;
    });

    html += '</tbody></table></div>';
    container.innerHTML = html;
}

// Create new data entry
async function createDataEntry() {
    const companySelect = document.getElementById('entry-company');
    if (!companySelect) {
        console.error('entry-company select not found');
        return;
    }
    
    const companyId = companySelect.value;
    if (!companyId) {
        showMessage('data-message', 'Please select a company', 'error');
        return;
    }

    // Get form data
    const entryData = {
        company_id: parseInt(companyId),
        device_type: document.getElementById('entry-device-type')?.value || '',
        uid: document.getElementById('entry-uid')?.value || '',
        data_type: document.getElementById('entry-data-type')?.value || '',
        data_set: document.getElementById('entry-data-set')?.value || '',
        data_going_to: document.getElementById('entry-data-going-to')?.value || ''
    };

    // Validate required fields
    if (!entryData.uid.trim()) {
        showMessage('data-message', 'UID is required', 'error');
        return;
    }

    try {
        await apiRequest('/data-entries', {
            method: 'POST',
            body: JSON.stringify(entryData)
        });
        
        // Clear form
        clearDataEntryForm();
        
        showMessage('data-message', 'Data entry created successfully');
        loadDataEntries();
    } catch (error) {
        showMessage('data-message', 'Failed to create data entry', 'error');
    }
}

// Clear data entry form
function clearDataEntryForm() {
    const formFields = [
        'entry-company',
        'entry-device-type',
        'entry-uid',
        'entry-data-type',
        'entry-data-set',
        'entry-data-going-to'
    ];
    
    formFields.forEach(fieldId => {
        const field = document.getElementById(fieldId);
        if (field) {
            field.value = '';
        }
    });
}

// Delete data entry
async function deleteDataEntry(id) {
    if (!confirm('Are you sure you want to delete this data entry?')) return;

    try {
        await apiRequest(`/data-entries/${id}`, { method: 'DELETE' });
        showMessage('data-message', 'Data entry deleted successfully');
        loadDataEntries();
    } catch (error) {
        showMessage('data-message', 'Failed to delete data entry', 'error');
    }
}

// Filter data entries
function filterDataEntries() {
    loadDataEntries(); // Reload with current filter values
}

// Clear filters
function clearFilters() {
    const filterFields = [
        'filter-company',
        'filter-uid', 
        'filter-data-set'
    ];
    
    filterFields.forEach(fieldId => {
        const field = document.getElementById(fieldId);
        if (field) {
            field.value = '';
        }
    });
    
    loadDataEntries(); // Reload without filters
}

async function uploadCSVFile(file) {
    const status = document.getElementById('upload-status');

    if (!file || file.type !== 'text/csv') {
        status.textContent = 'Please upload a valid CSV file.';
        return;
    }

    const formData = new FormData();
    formData.append('csv_file', file);

    status.textContent = 'Uploading...';

    fetch('/data-entries/upload-csv', {
        method: 'POST',
        body: formData
    }).then(res => res.json())
        .then(data => {
            if (data.error) {
                status.textContent = `Error: ${data.error}`;
            } else {
                status.textContent = `Upload successful: ${data.message}`;
                loadDataEntries();
            }
        }).catch(() => {
            status.textContent = 'Upload failed. Please try again.';
        });
}


function addDataEntryListener() {
    const dropArea = document.getElementById('csv-upload-area');
    const fileInput = document.getElementById('csv-file-input');

    dropArea.addEventListener('click', () => fileInput.click());

    ['dragenter', 'dragover'].forEach(eventName =>
        dropArea.addEventListener(eventName, e => {
            e.preventDefault();
            dropArea.classList.add('hover');
        })
    );

    ['dragleave', 'drop'].forEach(eventName =>
        dropArea.addEventListener(eventName, e => {
            e.preventDefault();
            dropArea.classList.remove('hover');
        })
    );

    dropArea.addEventListener('drop', e => {
        const file = e.dataTransfer.files[0];
        uploadCSVFile(file);
    });

    fileInput.addEventListener('change', () => {
        const file = fileInput.files[0];
        uploadCSVFile(file);
    });
}


// Event Listeners
if (document.getElementById("data-entries-page")) {
    document.addEventListener('DOMContentLoaded', () => {
        loadCompaniesForDataFilters();
        loadDataEntries();
        addDataEntryListener();
    });

    window.loadDataEntries = loadDataEntries;
    window.openEntryModal = openEntryModal;
    window.closeEntryModal = closeEntryModal;
    window.loadCompaniesForDataFilters = loadCompaniesForDataFilters;
    window.createDataEntry = createDataEntry;
    window.deleteDataEntry = deleteDataEntry;

}
